README.md
[![codecov](https://codecov.io/gh/leeprichman/ClonoCluster/branch/main/graph/badge.svg?token=GBJDQCGAWZ)](https://codecov.io/gh/leeprichman/ClonoCluster)![](https://img.shields.io/docker/pulls/leeprichman/clonocluster)
# ClonoCluster

```

#####                               #####                                           
#       #       ####  #    #  ####  #       #      #    #  ####  ##### ###### #####  
#       #      #    # ##   # #    # #       #      #    # #        #   #      #    #
#       #      #    # # #  # #    # #       #      #    #  ####    #   #####  #    #
#       #      #    # #  # # #    # #       #      #    #      #   #   #      #####  
#       #      #    # #   ## #    # #       #      #    #      #   #   #      #   #  
#####   ######  ####  #    #  ####   #####  ######  ####   ####    #   ###### #    #

```
*a method for using clonal origin to inform transcriptome clustering*

![](https://github.com/leeprichman/ClonoCluster/blob/main/readme_fig.png)

Welcome to ClonoCluster. To get started, you will need:

  1. A normalized or scaled count matrix of your barcoded cells

  2. A table of unique cell IDs and their assigned barcodes

You may also use sample data and a worked example demonstrated in the [tutorial](https://htmlpreview.github.io/?https://github.com/leeprichman/ClonoCluster/blob/main/Tutorial.html). If you are looking for the analysis and raw data from our 2022 paper, check out [this repo](https://github.com/leeprichman/ClonoCluster_paper).

## Dependencies

  * R version >= 3.6

  * `devtools` package for installation

## Installation

ClonoCluster is easily installed with the help of the `devtools` package:

```
install.packages("devtools")

devtools::install_github("leeprichman/ClonoCluster")
```

## Docker

ClonoCluster is also availble to use within a prebuilt docker image:

```
# pull the docker image
docker pull leeprichman/clonocluster

# launch a container

cID=$(docker run -it -d leeprichman/clonocluster /bin/bash)

# move any desired files onto the container

docker cp myfile.txt $cID:/myfilecopy.txt

# launch the interactive container

docker exec -it $cID bash

# launch R

R

```

Now you may follow the [tutorial](https://htmlpreview.github.io/?https://github.com/leeprichman/ClonoCluster/blob/main/Tutorial.html) or walkthrough below. At the end of the analysis you may recover output and shut down the container with:

```
docker cp $cID:/myoutput.txt ~/myagdockeroutput.txt

docker stop $cID

docker rm $cID

```

### Quickstart guide

#### Prep your input data

The input files you will need are:

 1. a two column table of cell IDs and Barcode assignments with the column names "rn" and "Barcode" respectively.

 2. a count matrix, with a a first column, "rn", of cell IDs and columns of gene counts

The latter can be retrieved from your Seurat object if you already have one:

```

cm <- myseuratobject[["RNA"]]@data

# transpose it so rows are cells and columns are genes
cm <- t(as.matrix(cm))

cm <- data.table::as.data.table(cm, keep.rownames = TRUE)

# save to file
data.table::fwrite(cm, "mycountmatrix.tsv", sep = "\t")

```

**N.B.:** *Make sure that cell IDs present in the barcode table are the only ones present in the count matrix and vice versa, otherwise you will get errors!*

#### ClonoCluster

For a worked example using provided sample data, check out the [tutorial](https://htmlpreview.github.io/?https://github.com/leeprichman/ClonoCluster/blob/main/Tutorial.html).

The first step is to generate our PCA matrix:

```
library(magrittr)
library(data.table)
library(ClonoCluster)

cm <- data.table::fread("mycountmatrix.tsv")

# convert data table to matrix
# make sure that rows are cells and columns are genes!
cm %<>% dt2m

# using 25 PCs
pca <- irlba_wrap(cm, npc = 25)

```

Now read in barcodes and ClonoCluster!

```
# two column table, "rn" (cell ID) and "Barcode"
bt <- data.table::fread("mybarcodetable.tsv")

# return the cluster assignments for range of alphas
clust <- clonocluster(pca, bt, alpha = seq(0, 1, by = 0.1), beta = 0.1, res = 1)

```

`clust` is a data table with the cluster assignment of each cell at all alpha values.

#### Sankey visualization

Let's make a Sankey plot of the data. We will use the top 10 barcodes so that we can take the plot all the way to alpha = 1 without being unable to interpret it.

```
# get top 10 barcodes
wl <- bt[, .N, by = "Barcode"] %>% .[order(-N), Barcode[1:10]]

# get the cell ids from the top 10 barcodes
wl <- bt[Barcode %chin% wl, rn]

# plot alluvia
Plot_alluvia(clust[rn %chin% wl], # subset on cell IDs
              bt[rn %chin% wl], # subset on cell IDs
              col = cw_colors, # provided vector of colors
              border_size = 0.5,
              label_nodes = FALSE,
              xlab = "\u03B1", # unicode for alpha
              ylab = "# of cells")

```

This is a sample of what this looks like for the top 10 barcodes in the tutorial sample data:

![](https://github.com/leeprichman/ClonoCluster/blob/main/sankey_sample.png)

#### UMAP and Warp Factor

You can also generate UMAPs projections for visualization:

```

# warp factor of 0 (default UMAP)
umap_wf0 <- engage_warp(pca, bt, 0)

# warp factor of 5
umap_wf5 <- engage_warp(pca, bt, 5)

# warp factor of 10 (maximum)
umap_wf10 <- engage_warp(pca, bt, 10)

# combine to one table
umap <- list(umap_wf0, umap_wf5, umap_wf10) %>% data.table::rbindlist()

# plot
ggplot(umap, aes(x = UMAP_1, UMAP_2)) +
  geom_point(col = "dodgerblue") +
  facet_wrap(~warp, nrow = 1) +
  theme_void()

```

![](https://github.com/leeprichman/ClonoCluster/blob/main/sample_warps.png)

#### Other functions

Check out the [tutorial](https://htmlpreview.github.io/?https://github.com/leeprichman/ClonoCluster/blob/main/Tutorial.html) for:

  * Cluster number analysis to choose alpha values

  * Barcode to cluster concordance plots with `cast_confusion`

  * Cluster marker analysis with `Find_Markers_ROC`

  * Marker fidelity Sankey plots with `Plot_alluvia_track`

  * Marker fidelity heatmaps with `pheatmap`

  * Overlay clusters and UMI counts on warped UMAPs

## Testing

ClonoCluster is supported by unit testing with `testthat` and [codecov.io](about.codecov.io). To test the package after install:

```
testthat::test_package("ClonoCluster")
```

## Citation

Richman *et al.* "ClonoCluster: a method for using clonal origin to inform transcriptome clustering" *biorXiv* 2022 [link](https://www.biorxiv.org/content/10.1101/2022.02.11.480077v1?rss=1)

## Selected references

  * Wagner and Klein "Lineage tracing meets single-cell omics: opportunities and challenges" *Nature Reviews Genetics* 2020 [link](https://www.nature.com/articles/s41576-020-0223-2)

  * Goyal *et al.* “Cell Type Determination for Cardiac Differentiation Occurs Soon after Seeding of Human Induced Pluripotent Stem Cells.” *bioRxiv* 2021 [link](https://doi.org/10.1101/2021.08.08.455532)

## Contact

Open an issue on this GitHub repository, contact [@leeprichman](https://github.com/leeprichman), [@arjunrajlab](https://github.com/arjunrajlab), or email [Lee Richman](leeprichman@gmail.com) or [Dr. Arjun Raj](arjunrajlaboratory@gmail.com).


NEWS.md
# ClonoCluster v0.0.1


readme_figs.R
library(data.table)
library(magrittr)
library(Seurat)
library(ClonoCluster)
library(ggplot2)
library(pheatmap)
library(cowplot)

# get the location of the sample data installed with ClonoCluster
dir <- system.file(package = "ClonoCluster") %>% file.path(., "extdata")

# count matrix file
cm <- file.path(dir, "YG1_sample_genes.txt") %>% data.table::fread() %>% tdt %>% dt2m

# barcode assignment file
bt <- file.path(dir, "YG1_sample_barcodes.txt") %>% data.table::fread()

# sankets
pca <- irlba_wrap(cm, npc = 25)

als <- seq(0, 1, by = 0.1)

# return the cluster assignments for range of alphas
clust <- ClonoCluster(pca, bt, alpha = als, beta = 0.1, res = 1.5)

wl <- bt[, .N, by = "Barcode"] %>% .[order(-N), Barcode[1:10]]

wl <- bt[Barcode %chin% wl, rn]

p1 <- Plot_alluvia(clust[rn %chin% wl & alpha < 0.6],
                  bt[rn %chin% wl],
                  label_nodes = FALSE,
                  reverse = FALSE,
                  border_size = 0.5,
                  col = cw_colors
                  ) +
                  theme_void() +
                  theme(legend.position = "none",
                  plot.title = element_blank(),
                  plot.subtitle = element_blank())

ggsave(plot = p1, "sankey_void.pdf", height = 3, width = 6)

p2 <- Plot_alluvia(clust[rn %chin% wl],
                  bt[rn %chin% wl],
                  col = cw_colors,
                  border_size = 0.5,
                  label_nodes = FALSE,
                  xlab = "\u03B1",
                  ylab = "# of cells")

ggsave(plot = cowplot::plot_grid(plotlist = p2, ncol = 1), "sankey_sample.png", height = 6, width = 6)

# warp factor of 0 (default UMAP)
umap_wf0 <- engage_warp(pca, bt, 0)

# warp factor of 5
umap_wf5 <- engage_warp(pca, bt, 5)

# warp factor of 10 (maximum)
umap_wf10 <- engage_warp(pca, bt, 10)

# combine to one table
umap <- list(umap_wf0, umap_wf5, umap_wf10) %>% data.table::rbindlist()

umap %<>% .[order(warp)]

umap[, warp := paste("Warp factor =", warp)]

umap[, warpf := factor(warp,
  levels = c("Warp factor = 0", "Warp factor = 5", "Warp factor = 10"),
  ordered = TRUE)]

# plot
p <- ggplot(umap, aes(x = UMAP_1, UMAP_2)) +
  geom_point(col = cw_colors[2], size = 0.1, alpha = 0.4) +
  facet_wrap(~warpf, nrow = 1) +
  theme_void()

ggsave(plot = p, "sample_warps.png", height = 2, width = 6)

# warps for svg
# lets do multiple warp factors
wfs <- c(0, 4, 8)

# get our warped UMAPs
umaps <- lapply(wfs, function(s){

  uws <- engage_warp(pca, bt, s)

  return(uws)

}) %>% data.table::rbindlist()

# add our barcodes to the table

umaps <- merge(umaps, bt, by = "rn")

# color by barcode and put singlets into one category
umaps[, Barcode :=
        ifelse(rn %>% unique %>% length > 1, Barcode, "Singlet"),
        by = "Barcode"]

ums <- ggplot(umaps[Barcode != "Singlet"], aes(x = UMAP_1, y = UMAP_2)) +
    geom_point(aes(col = Barcode), size = 0.1, alpha = 0.5) +
    facet_wrap(~warp) +
    scale_color_manual(values = c(cw_colors, c25)) +
    ttheme +
    theme_void() +
    theme(legend.position = "none", strip.text = element_blank())

ggsave(plot = ums, "warps_void.pdf", height = 2, width = 6)
